src/ctia/schemas/utils.clj

```diff
@@ -18,55 +19,169 @@
-          (s/specific-key? k) (f (s/explicit-schema-key k))
-          :else k))
-      schema))))
-
+;; TODO delete me
 (defn optional-keys
   "Makes given map keys optional. Defaults to all keys."
-  ([m] (optional-keys m nil))
-  ([m ks] (transform-keys m s/optional-key ks)))
+  ([m] (st/optional-keys m))
+  ([m ks] (st/optional-keys m ks)))
 
+;; TODO delete me
 (defn optional-keys-schema
   "Walks a schema making all keys optional in Map Schemas."
   [schema]
-  (sw/prewalk
-   (fn [x]
-     (if (and (map? x) (not (record? x)))
-       (optional-keys x)
-       x))
-   schema))
+  (st/optional-keys-schema schema))
+
+(s/defschema SpecificKey
+  (s/pred s/specific-key?))
+
+(s/defn service-subgraph :- (s/pred map?)
+  "Returns a subgraph of a Trapperkeeper service graph value.
+  If any required levels are missing, throws an exception.
+  
+  (service-subgraph
+    {:ConfigService {:get-in-config <...>
+                     :get-config <...>}
+     :FooService {:f1 <...>
+                  :f2 <...>}
+     :BarService {:b1 <...>}}
+    {:ConfigService #{:get-config}
+     :FooService #{:f2 (s/optional-key :f3)}
+     (s/optional-key :MissingService) #{:m1}})
+  ;=> {:ConfigService {:get-config <...>}
+  ;    :FooService {:f2 <...>}}
+  "
+  [graph-value :- (s/pred map?)
+   selectors :- {SpecificKey #{SpecificKey}}]
+  (into {}
+        (map (fn [[service-kw fn-kws]]
+               (assert (s/specific-key? service-kw)
+                       (pr-str service-kw))
+               (assert (simple-keyword? (s/explicit-schema-key service-kw))
+                       (s/explicit-schema-key service-kw))
+               (assert (set? fn-kws))
+               (let [gval (get graph-value (s/explicit-schema-key service-kw))]
+                 (if-not (map? gval)
+                   (when-not (s/optional-key? service-kw)
+                     (throw (ex-info (str "Missing service: " service-kw) {})))
+                   (let [service-fns (into {}
+                                           (map (fn [fn-kw]
+                                                  (assert (simple-keyword? (s/explicit-schema-key fn-kw))
+                                                          (s/explicit-schema-key fn-kw))
+                                                  (if-some [svc-fn (get gval (s/explicit-schema-key fn-kw))]
+                                                    {(s/explicit-schema-key fn-kw) svc-fn}
+                                                    (when-not (s/optional-key? fn-kw)
+                                                      (throw (ex-info (str "Missing " service-kw " service function: "
+                                                                           (s/explicit-schema-key fn-kw))
+                                                                      {}))))))
+                                           fn-kws)]
+                     {(s/explicit-schema-key service-kw) service-fns})))))
+        selectors))
```

Comment:
I am somehow lost in this conditional tree :-)
Could you split it a little bit to ease its reading.

Reply:
I named the conditions and added a comment, does that help?